### ISSUE TITLE
"parlai train_model" CLI fails in python 3.8

### ISSUE THREAD

**Issue by lazyrenee** [facebookresearch/ParlAI#3119]
**Bug description**
`parlai train_model` for `parlai 0.9.2` seems not working. 

**Reproduction steps**
I tried to run `parlai train_model -t twitter -mf /tmp/tr_twitter -m transformer/ranker` as the first example to test after I just rebased the repository, it went wrong. I checked with the the current parlai version is `parlai 0.9.2` But this doesn't happen to my previous one `parlai 0.9`, not sure if that's something related to version update.

**Logs**
Please paste the command line output:
```
(base) xxx$ parlai train_model -t twitter -mf /tmp/tr_twitter -m transformer/ranker
usage: parlai build_candidates [-h] [--helpall] [-o INIT_OPT] [-t TASK] [-dt DATATYPE] [-bs BATCHSIZE]
                               [-dynb {full,batchsort,None}] [-dp DATAPATH] [-n NUM_EXAMPLES] [-of OUTFILE]
                               [-ltim LOG_EVERY_N_SECS]

Build the candidate responses for a retrieval model

optional arguments:
  -h, --help
        show this help message and exit
  --helpall
        Show usage, including advanced arguments.
  -o, --init-opt INIT_OPT
        Path to json file of options. Note: Further Command-line arguments override file-based options. (default: None)
  -t, --task TASK
        ParlAI task(s), e.g. "babi:Task1" or "babi,cbt" (default: None)
  -dt, --datatype DATATYPE
        choose from: train, train:ordered, valid, test. to stream data add ":stream" to any option (e.g.,
        train:stream). by default train is random with replacement, valid is ordered, test is ordered. (default:
        train:evalmode)
  -bs, --batchsize BATCHSIZE
        batch size for minibatch training schemes (default: 1)
  -dynb, --dynamic-batching {full,batchsort,None}
        Use dynamic batching (default: None)
  -dp, --datapath DATAPATH
        path to datasets, defaults to {parlai_dir}/data (default: None)
  -n, --num-examples NUM_EXAMPLES
        Total number of exs to convert, -1 to convert all examples (default: -1)
  -of, --outfile OUTFILE
        Output file where to save, by default will be created in /tmp (default: None)
  -ltim, --log-every-n-secs LOG_EVERY_N_SECS

optional arguments:
  -h, --help
        show this help message and exit
  --helpall
        Show usage, including advanced arguments.
  -n, --num-examples NUM_EXAMPLES
        Total number of exs to convert, -1 to convert all examples (default: -1)
  -of, --outfile OUTFILE
        Output file where to save, by default will be created in /tmp (default: None)
  -ltim, --log-every-n-secs LOG_EVERY_N_SECS

Main ParlAI Arguments:
  -o, --init-opt INIT_OPT
        Path to json file of options. Note: Further Command-line arguments override file-based options. (default: None)
  -t, --task TASK
        ParlAI task(s), e.g. "babi:Task1" or "babi,cbt" (default: None)
  -dt, --datatype DATATYPE
        choose from: train, train:ordered, valid, test. to stream data add ":stream" to any option (e.g.,
        train:stream). by default train is random with replacement, valid is ordered, test is ordered. (default:
        train:evalmode)
  -bs, --batchsize BATCHSIZE
        batch size for minibatch training schemes (default: 1)
  -dynb, --dynamic-batching {full,batchsort,None}
        Use dynamic batching (default: None)
  -dp, --datapath DATAPATH
        path to datasets, defaults to {parlai_dir}/data (default: None)

Parse Error: argument -mtw/--multitask-weights: invalid 'multitask_weights' value: 'transformer/ranker'
```

**Additional context**
Add any other context about the problem here. (like proxy settings, network setup, overall goals, etc.)

**Comment by stephenroller** [comment]
Ugh yes, this is a bug with python 3.8 and parlai I haven't figured out yet.

Use --model instead of -m and --task instead of -t as a temporary mitigation.

**Comment by stephenroller** [comment]
Alternatively, if python 3.7 is viable, you could use it.

**Comment by lazyrenee** [comment]
Thanks!!! so much appreciated! will try :)

**Comment by stephenroller** [comment]
Reopening to track the root cause

**Comment by lazyrenee** [comment]
uptate: I changed my current python version to `python 3.7.9` but seems still not working for `parlai train_model -t twitter -mf /tmp/tr_twitter -m transformer/ranker` command 
```
(base) xxx$ python --version
Python 3.7.9
```

**Comment by lazyrenee** [comment]
I also tried to reinstall using the command `python setup.py develop` after I downgraded to `python 3.7` 

The log shows:
```
(base) xxx:~/workspace/ParlAI$ python setup.py develop
/usr/lib/python3.7/distutils/dist.py:274: UserWarning: Unknown distribution option: 'long_description_content_type'
  warnings.warn(msg)
/usr/lib/python3.7/distutils/dist.py:274: UserWarning: Unknown distribution option: 'python_requires'
  warnings.warn(msg)
running develop
Checking .pth file support in /usr/local/lib/python3.7/dist-packages/
error: can't create or remove files in install directory

The following error occurred while trying to add or remove files in the
installation directory:

    [Errno 13] Permission denied: '/usr/local/lib/python3.7/dist-packages/test-easy-install-18365.pth'

The installation directory you specified (via --install-dir, --prefix, or
the distutils default setting) was:

    /usr/local/lib/python3.7/dist-packages/

Perhaps your account does not have write access to this directory?  If the
installation directory is a system-owned directory, you may need to sign in
as the administrator or "root" account.  If you do not have administrative
access to this machine, you may wish to choose a different installation
directory, preferably one that is listed in your PYTHONPATH environment
variable.

For information on other options, you may wish to consult the
documentation at:

  https://pythonhosted.org/setuptools/easy_install.html

Please make the appropriate changes for your system and try again.
```

Then I tried using `easy_install` as well as `sudo python setup.py develop` The logs are below:
```
(base) xxxx:~/workspace/ParlAI$ python --version
Python 3.7.9
(base) eilab@courage:~/workspace/ParlAI$ sudo python setup.py develop
Sorry, Python >=3.6 is required for ParlAI.
(base) xxxx:~/workspace/ParlAI$ easy_install setup.py develop
WARNING: The easy_install command is deprecated and will be removed in a future version.
Processing setup.py
error: Couldn't find a setup script in /home/workspace/ParlAI/setup.py
```

**Comment by stephenroller** [comment]
For context, https://github.com/python/cpython/commit/b1e4d1b6032d4c82b549233fa08a2c7cfe7e818b is where the short options were broken in python 3.8.

**Comment by stephenroller** [comment]
@lazyrenee for that task, you need to use `--candidates batch --eval-candidates batch`.

**Comment by lazyrenee** [comment]
I see, thanks for getting back so quickly :) For now guess I'll just continue to use `python 3.7` for the parlai CLIs I'm more familiar with.

**Comment by stephenroller** [comment]
I'm thinking about how to solve this problem. It looks to me like python has straight banned the behavior we were using.

The only remedy I can think of is switching to `--mtw` to replace `-mtw`.

**Comment by lazyrenee** [comment]
Hi @stephenroller , 

Thanks for getting back to me so quick :) 

I checked the code, it seems the issue with `argparse` and `python 3.8`. I tried to explore a bit and this seems the issue with the argparse & `add_help` parameter, so I wrote a tiny script to test it out: 

```
import argparse
parser = argparse.ArgumentParser(description = 'tests.', add_help = True)

parser.add_argument('-s', '--short', type = str)

args = parser.parse_args()
print("test for -s with args.short: {}".format(args.short))
print("test for -short with args.short: {}".format(args.short))
```

The test result is as follows:
```
(python3.8) xxxx:~$ conda activate python3.8
(python3.8) xxxx:~$ python --version
Python 3.8.3
(python3.8) xxxx:~$ python exp_args.py --short Hello_world
test for -s with args.short: Hello_world
test for -short with args.short: Hello_world
(python3.8) xxxx:~$ python exp_args.py -s Hello_world
test for -s with args.short: Hello_world
test for -short with args.short: Hello_world
```
may could enable to : https://github.com/facebookresearch/ParlAI/blob/master/parlai/core/params.py#L283

(not fully sure if that could solve the problem)

Best,
Renee

**Comment by stephenroller** [comment]
This was resolved with #3284
